### DISCUSSION TITLE
TheXTech Port Possibility List (Unofficial)

### DISCUSSION THREAD

**Discussion by ghost** [Wohlstand/TheXTech#685] [Planning]
# What is already ported:

- Windows NT after 2000 (XP, Vista, 7, 8, 8.1, 10, 11)
- Apple macOS (10.4/Tiger up to the latest macOS release, 10.23/Sonoma)
- Linux (Any distribution)
- Android (Is there a minimum version?)
- Emscripten
- Nintendo Wii (Hacked Wii only)
- Nintendo 3DS (Hacked 3DS only, a 3DSX file is only provided. A CIA file can increase the power TheXTech can use, especially on a N3DS, but the developers chose not to generate CIA files for the game)
- Nintendo Wii U (Recent system update with a hacked Wii U will work, I don't know if older Wii U versions will work)
- Nintendo Switch (Hacked Switch only)
- Sony Playstation Vita (Hacked system only)

# What can be possible to port it to:

- Apple iOS (iPhone, iPad, iPod): Though it can be possible as a sideloaded app/jailbroken tweak app, Wohlstand does not have any iOS devices in order to support/test TheXTech on
- Windows 2000: This will need some WinAPI calls either edited out or have some replacements to the calls used from XP and above, to fully support Windows 2000 (dbghelp.dll is one file not available on Windows 2000)
- Windows 98: The SDL fork by Wohlstand will need to have 9x support before it can be possible to port
- Nintendo GameCube: Less RAM than the Wii, but can be possible. SD Card access needs to be possible (Memory Card with a SD Card slot, GC Loader, FlippyDrive, SD2SP2) to load assets for TheXTech, or the game needs to store every assets into the game itself
- Sony Playstation 2: Can be possible (USBs can be used but it's limited to USB 1.1 speeds)
- Sony Playstation 3-5: Is powerful enough to support. As for PS5, there isn't much hacked systems for that console, but still
- Microsoft Xbox, Original Edition: Can be possible to support but with caveats like a worse user experience
- Microsoft Xbox 360, One, Series X/S: Is powerful enough to support. 360 can only run TheXTech with hacked 360's, while One and above can use Developer Mode to run the game
- Nintendo DSi: It has 16 MBs of RAM, unlike the original Nintendo DS, and has an SD Card slot. It might be possible to port to, as long as someone experienced comes in to port it to the DSi
- Sega Dreamcast: Can be possible to support but with caveats like a worse user experience
- Apple macOS 10.0-10.3: Issue relies on a lack of a C++11 toolchain, but if that gets resolved then it'll be possible
- Apple Mac OS 9 (Classic): Pre-Mac OS X. There's a lack of SDL, and a lack of a C++11 toolchain, but can still be possible if those things get bypassed

# What can't be possible to port to:

- Windows 95: SDL1 doesn't have native 95 support
- Windows 3.1 and below
- Sony Playstation 1: It doesn't have a way to read assets externally, but storing assets in-game could be. Though this system is probably not possible due to it being weaker than the PS2
- Nintendo DS: There's barely any RAM, unlike the DSi, and if not running on a flashcart it'll be tricky to fit every asset into a NDS file...
- Nintendo Game Boy Advance: Unless they pull an OpenLara, it won't be possible
- Nintendo Game Boy
- Nintendo Game Boy Color
- Nintendo 64: Max RAM is 8 MBs from the Expansion Pak. TheXTech needs at least 16 MBs to run
- Super Nintendo Entertainment System: What are you, stupid? SNES games run in assembly, not C++
- Nintendo Entertainment System: Same as SNES (But not as powerful as the SNES)
- Sega SG-1000
- Sega Master System
- Sega Genesis
- Sega CD
- Sega Pico
- Sega Saturn
- Sega 32X: Unless they pull an OpenLara, it won't be possible
- Atari 2600: Do I seriously need to explain this one?
- Atari 5200
- Atari 7600
- Atari ST
- Atari Lynx
- Atari Jaguar
- Atari Jaguar CD
- Game.com
- Google Stadia: If you were under a rock for so many years, it shut down. Also it was an online gaming service, not anything with Homebrew development
- ZX Spectrum
- V.Smile

Last updated on January 14th, 2024

**Comment by ds-sloth** [comment]
Very nice summary!

Some notes regarding work that can be done:
- Apple macOS: it would be nice to know exactly what the minimum supported version is for the CI builds.
- Mac OS X: there isn't currently a CI for the legacy 10.4 port, and I don't know whether the OpenGL renderer currently functions for it. It's extremely unlikely we'd get OpenGL shader support there (this requires OpenGL 3+), but we could at least get faster rendering and logic op support. 10.0-10.3 should also be possible to port to. The main issue is the lack of a well-supported C++11 toolchain for versions earlier than Tiger.
- Mac OS Classic: there isn't SDL support here, but we don't need it for a functioning port if someone is motivated enough. Mac OS 9.0 includes OpenGL 1.2 and it's possible that Mac OS 8.6 has OpenGL 1.1 support (I'm not sure). The biggest issue will be getting a C++11 toolchain working, and the second big issue will be figuring out audio support so that MixerX can run.
- Windows 95/98: if a computer does not meet the minimum CPU / RAM requirements for Windows 2000, it won't run an SDL port of TheXTech smoothly, and even a custom port might not have a good user experience.
- Apple iOS: this should be fully supported by Emscripten (and TheXTech is a "Progressive Web App" so it supports offline play) but there are some frame timing issues and occasional crashes due to mobile Safari having lower limits than other browsers

All fifth-generation and earlier consoles, and the handheld equivalents (including GBA and Nintendo DS) are not realistic due to CPU and/or RAM constraints.

The PSP, DSi, XBox, PS2, and Dreamcast are probably possible with a motivated enough developer but will likely have a worse user experience than the Wii port, which is currently our weakest port. That said, the biggest obstacle to a smooth user experience (even at this level of CPU / RAM) is storage read speed, so who knows?

**Reply by Wohlstand** [reply]
> lack of a well-supported C++11 toolchain for versions earlier than Tiger

Just because Tiger has a custom Tiger-Brew toolchain that gives the GCC 7 with C++11 and other nifty things. Additionally, older Mac OS X do lack a lot of API functions and it's will be tricky to support them. The fact, I had to seriously hack the SDL2 to make it work on Tiger (the SDL2 itself supports 10.7 as the minimal OS), however, it's not completed: I should complete the support for Cocoa video, right now the thing runs the X11 which just works, but it isn't convenient for people as it requires to manually run the X11 server to run the thing.

**Reply by Wohlstand** [reply]
> Apple iOS: this should be fully supported by Emscripten (and TheXTech is a "Progressive Web App" so it supports offline play) but there are some frame timing issues and occasional crashes due to mobile Safari having lower limits than other browsers

Also, there is a large sound latency which may absent if run as native app. In theory, it's possible to make a native port that could work via side-loading or on jail-broken iOS devices, but there are problems right now:
- I don't own any iOS devices at all: only simulators.
- I don't have enough knowledge on how to deal with several systems of the iOS device like file systems, etc. Other rest could just work through SDL2's functions.

**Comment by ds-sloth** [comment]
Ah, one note: this discussion duplicates #651 to some extent.

**Comment by Wohlstand** [comment]
Some notes from me:

> Windows NT (2000, XP, Vista, 7, 8, 8.1, 10, 11)

Windows XP is minimal right now, on Windows 2000 it won't start. It uses WinAPI functions are unavailable in 2000, example: 
![Снимок экрана от 2024-01-09 02-16-42](https://github.com/Wohlstand/TheXTech/assets/6751442/9ee88cc0-a819-466d-8f6b-0c5c22d9575b)

> Windows 95/Windows 98: The SDL fork by Wohlstand will need to have 9x support before it can be possible to port

The Windows 98 could be, but to 95 - very doubtable :thinking:. The fact: an SDL1 has native Windows 98 support, but SDL2 doesn't. On my end, I started an experiment to make an SDL2 port for 9x, but it didn't finished, I made it load a simple demo, however, it fails to initialize the video for unknown reasons, and I didn't finished it because I became busy on other things. Additionally, I should also implement a kin dof a workaround for the Unicode thing as 9x totally lacks it. There are some libraries do exist, but they doesn't work as intended (I attempted). Additonally, it's impossible to have modern C++ standards on MSVC and support old Windows at the same time. MinGW and MinGW-w64 allows that (with GCC up to 6, the newer GCC and libwinpthread requires a patching).

**Comment by ghost** [comment]
I updated the Port List to have a better list of what can be ported and can't, thanks to the devs specifying what can be possible and what isn't.

**Reply by Wohlstand** [reply]
Android minimum version is 4.1.

**Comment by vitzie629** [comment]
Add Sony PlayStation portable support for TheXTech, regardless of whether the PSP is hacked or not and the PSP model variations don’t matter at all, which means it would work with all PSP models!

**Comment by LoveBodhi** [comment]
My other suggestion is HarmonyOS NEXT support.
